<a href="https://colab.research.google.com/github/wandb/examples/blob/master/colabs/pyg/8_Node_Classification_(with_W&B).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>
<!--- @wandbcode{pytorch_geometric_example} -->

<img src="http://wandb.me/logo-im-png" width="400" alt="Weights & Biases" />
<!--- @wandbcode{intro-colab} -->

In [ ]:
enable_wandb = True

if enable_wandb:
  !pip install -qqq wandb
  import wandb
  wandb.login()

     |████████████████████████████████| 1.8 MB 10.7 MB/s 
     |████████████████████████████████| 181 kB 27.3 MB/s 
     |████████████████████████████████| 147 kB 37.4 MB/s 
     |████████████████████████████████| 63 kB 971 kB/s 


<IPython.core.display.Javascript object>

wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter, or press ctrl+c to quit: 

··········


wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc


In [ ]:
# Install required packages.
import os
import pdb
import torch
import pandas

os.environ['TORCH'] = torch.__version__
print(torch.__version__)

!pip install -q torch-scatter -f https://data.pyg.org/whl/torch-${TORCH}.html
!pip install -q torch-sparse -f https://data.pyg.org/whl/torch-${TORCH}.html
!pip install -q git+https://github.com/pyg-team/pytorch_geometric.git

# Helper function for visualization.
%matplotlib inline
import matplotlib.pyplot as plt
from sklearn.manifold import TSNE

def visualize(h, color):
    z = TSNE(n_components=2).fit_transform(h.detach().cpu().numpy())
    plt.figure(figsize=(10,10))
    plt.xticks([])
    plt.yticks([])
    plt.scatter(z[:, 0], z[:, 1], s=70, c=color, cmap="Set2")
    plt.show()

def embedding_to_wandb(h, color, key="embedding"):
    num_components = h.shape[-1]
    df = pandas.DataFrame(data=h.detach().cpu().numpy(),
                        columns=[f"c_{i}" for i in range(num_components)])
    df["target"] = color.detach().cpu().numpy().astype("str")
    cols = df.columns.tolist()
    df = df[cols[-1:] + cols[:-1]]
    wandb.log({key: df})

1.12.0+cu113
     |████████████████████████████████| 7.9 MB 15.2 MB/s 
     |████████████████████████████████| 3.5 MB 9.2 MB/s 


# Node Classification with Graph Neural Networks

[Previous: Introduction: Hands-on Graph Neural Networks](https://colab.research.google.com/drive/1h3-vJGRVloF5zStxL5I0rSy4ZUPNsjy8)

This tutorial will teach you how to apply **Graph Neural Networks (GNNs) to the task of node classification**.
Here, we are given the ground-truth labels of only a small subset of nodes, and want to infer the labels for all the remaining nodes (*transductive learning*).

To demonstrate, we make use of the `Cora` dataset, which is a **citation network** where nodes represent documents.
Each node is described by a 1433-dimensional bag-of-words feature vector.
Two documents are connected if there exists a citation link between them.
The task is to infer the category of each document (7 in total).

This dataset was first introduced by [Yang et al. (2016)](https://arxiv.org/abs/1603.08861) as one of the datasets of the `Planetoid` benchmark suite.
We again can make use [PyTorch Geometric](https://github.com/rusty1s/pytorch_geometric) for an easy access to this dataset via [`torch_geometric.datasets.Planetoid`](https://pytorch-geometric.readthedocs.io/en/latest/modules/datasets.html#torch_geometric.datasets.Planetoid):

In [ ]:
from torch_geometric.datasets import Planetoid
from torch_geometric.transforms import NormalizeFeatures



dataset = Planetoid(root='data/Planetoid', name='Cora', transform=NormalizeFeatures())

print()
print(f'Dataset: {dataset}:')
print('======================')
print(f'Number of graphs: {len(dataset)}')
print(f'Number of features: {dataset.num_features}')
print(f'Number of classes: {dataset.num_classes}')

data = dataset[0]  # Get the first graph object.

print()
print(data)
print('===========================================================================================================')

# Gather some statistics about the graph.
print(f'Number of nodes: {data.num_nodes}')
print(f'Number of edges: {data.num_edges}')
print(f'Average node degree: {data.num_edges / data.num_nodes:.2f}')
print(f'Number of training nodes: {data.train_mask.sum()}')
print(f'Training node label rate: {int(data.train_mask.sum()) / data.num_nodes:.2f}')
print(f'Has isolated nodes: {data.has_isolated_nodes()}')
print(f'Has self-loops: {data.has_self_loops()}')
print(f'Is undirected: {data.is_undirected()}')

Processing...



Dataset: Cora():
Number of graphs: 1
Number of features: 1433
Number of classes: 7

Data(x=[2708, 1433], edge_index=[2, 10556], y=[2708], train_mask=[2708], val_mask=[2708], test_mask=[2708])
Number of nodes: 2708
Number of edges: 10556
Average node degree: 3.90
Number of training nodes: 140
Training node label rate: 0.05
Has isolated nodes: False
Has self-loops: False
Is undirected: True


Done!


Overall, this dataset is quite similar to the previously used [`KarateClub`](https://pytorch-geometric.readthedocs.io/en/latest/modules/datasets.html#torch_geometric.datasets.KarateClub) network.
We can see that the `Cora` network holds 2,708 nodes and 10,556 edges, resulting in an average node degree of 3.9.
For training this dataset, we are given the ground-truth categories of 140 nodes (20 for each class).
This results in a training node label rate of only 5%.

In contrast to `KarateClub`, this graph holds the additional attributes `val_mask` and `test_mask`, which denotes which nodes should be used for validation and testing.
Furthermore, we make use of **[data transformations](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#data-transforms) via `transform=NormalizeFeatures()`**.
Transforms can be used to modify your input data before inputting them into a neural network, *e.g.*, for normalization or data augmentation.
Here, we [row-normalize](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html#torch_geometric.transforms.NormalizeFeatures) the bag-of-words input feature vectors.

We can further see that this network is undirected, and that there exists no isolated nodes (each document has at least one citation).

## Training a Multi-layer Perception Network (MLP)

In theory, we should be able to infer the category of a document solely based on its content, *i.e.* its bag-of-words feature representation, without taking any relational information into account.

Let's verify that by constructing a simple MLP that solely operates on input node features (using shared weights across all nodes):

In [ ]:
import torch
from torch.nn import Linear
import torch.nn.functional as F


class MLP(torch.nn.Module):
    def __init__(self, hidden_channels):
        super().__init__()
        torch.manual_seed(12345)
        self.lin1 = Linear(dataset.num_features, hidden_channels)
        self.lin2 = Linear(hidden_channels, dataset.num_classes)

    def forward(self, x):
        x = self.lin1(x)
        x = x.relu()
        x = F.dropout(x, p=0.5, training=self.training)
        x = self.lin2(x)
        return x

model = MLP(hidden_channels=16)
print(model)

MLP(
  (lin1): Linear(in_features=1433, out_features=16, bias=True)
  (lin2): Linear(in_features=16, out_features=7, bias=True)
)


(optionally) logging the data attributes to W&B summary.

In [ ]:
if enable_wandb:
    wandb.init(project='node-classification')
    summary = dict()
    summary["data"] = dict()
    summary["data"]["num_features"] = dataset.num_features
    summary["data"]["num_classes"] = dataset.num_classes
    summary["data"]["num_nodes"] = data.num_nodes
    summary["data"]["num_edges"] = data.num_edges 
    summary["data"]["has_isolated_nodes"] = data.has_isolated_nodes()
    summary["data"]["has_self_nodes"] = data.has_self_loops()
    summary["data"]["is_undirected"] = data.is_undirected()
    summary["data"]["num_training_nodes"] = data.train_mask.sum()
    wandb.summary = summary

wandb: Currently logged in as: morg. Use `wandb login --relogin` to force relogin


Our MLP is defined by two linear layers and enhanced by [ReLU](https://pytorch.org/docs/stable/generated/torch.nn.ReLU.html?highlight=relu#torch.nn.ReLU) non-linearity and [dropout](https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html?highlight=dropout#torch.nn.Dropout).
Here, we first reduce the 1433-dimensional feature vector to a low-dimensional embedding (`hidden_channels=16`), while the second linear layer acts as a classifier that should map each low-dimensional node embedding to one of the 7 classes.

Let's train our simple MLP by following a similar procedure as described in [the first part of this tutorial](https://colab.research.google.com/drive/1h3-vJGRVloF5zStxL5I0rSy4ZUPNsjy8).
We again make use of the **cross entropy loss** and **Adam optimizer**.
This time, we also define a **`test` function** to evaluate how well our final model performs on the test node set (which labels have not been observed during training).

We also visualize the embeddings of the untrained model to in visually comparing the progress made by the training process below.


**NOTE**: *For W&B mode, please set up the embedding projector from the setting panel of the logged table. More information can be found here: https://docs.wandb.ai/ref/app/features/panels/weave/embedding-projector*

In [ ]:
from IPython.display import Javascript  # Restrict height of output cell.
display(Javascript('''google.colab.output.setIframeHeight(0, true, {maxHeight: 300})'''))

model = MLP(hidden_channels=16)

with torch.no_grad():
  out = model(data.x)

if enable_wandb:
    embedding_to_wandb(out, color=data.y, key="mlp/embedding/init")
else:
    visualize(out, data.y)

criterion = torch.nn.CrossEntropyLoss()  # Define loss criterion.
optimizer = torch.optim.Adam(model.parameters(), lr=0.01, weight_decay=5e-4)  # Define optimizer.

def train():
      model.train()
      optimizer.zero_grad()  # Clear gradients.
      out = model(data.x)  # Perform a single forward pass.
      loss = criterion(out[data.train_mask], data.y[data.train_mask])  # Compute the loss solely based on the training nodes.
      loss.backward()  # Derive gradients.
      optimizer.step()  # Update parameters based on gradients.
      return loss

def test():
      model.eval()
      out = model(data.x)
      pred = out.argmax(dim=1)  # Use the class with highest probability.
      test_correct = pred[data.test_mask] == data.y[data.test_mask]  # Check against ground-truth labels.
      test_acc = int(test_correct.sum()) / int(data.test_mask.sum())  # Derive ratio of correct predictions.
      return test_acc

for epoch in range(1, 201):
    loss = train()
    if enable_wandb:
        wandb.log({"mlp/loss": loss})
    print(f'Epoch: {epoch:03d}, Loss: {loss:.4f}')

<IPython.core.display.Javascript object>

Epoch: 001, Loss: 1.9616
Epoch: 002, Loss: 1.9564
Epoch: 003, Loss: 1.9495
Epoch: 004, Loss: 1.9419
Epoch: 005, Loss: 1.9369
Epoch: 006, Loss: 1.9244
Epoch: 007, Loss: 1.9196
Epoch: 008, Loss: 1.9131
Epoch: 009, Loss: 1.9019
Epoch: 010, Loss: 1.8895
Epoch: 011, Loss: 1.8737
Epoch: 012, Loss: 1.8616
Epoch: 013, Loss: 1.8539
Epoch: 014, Loss: 1.8458
Epoch: 015, Loss: 1.8197
Epoch: 016, Loss: 1.8194
Epoch: 017, Loss: 1.8027
Epoch: 018, Loss: 1.7732
Epoch: 019, Loss: 1.7733
Epoch: 020, Loss: 1.7491
Epoch: 021, Loss: 1.7472
Epoch: 022, Loss: 1.7194
Epoch: 023, Loss: 1.6962
Epoch: 024, Loss: 1.6584
Epoch: 025, Loss: 1.6668
Epoch: 026, Loss: 1.6226
Epoch: 027, Loss: 1.6333
Epoch: 028, Loss: 1.6029
Epoch: 029, Loss: 1.5534
Epoch: 030, Loss: 1.5783
Epoch: 031, Loss: 1.5381
Epoch: 032, Loss: 1.5213
Epoch: 033, Loss: 1.4926
Epoch: 034, Loss: 1.4652
Epoch: 035, Loss: 1.4754
Epoch: 036, Loss: 1.4559
Epoch: 037, Loss: 1.3932
Epoch: 038, Loss: 1.3604
Epoch: 039, Loss: 1.3321
Epoch: 040, Loss: 1.3120


After training the model, we can call the `test` function to see how well our model performs on unseen labels.
Here, we are interested in the accuracy of the model, *i.e.*, the ratio of correctly classified nodes:

We also visualize the embeddings of the output. This will give us a visual hint as to how good the model is performing, when compared to the embeddings of the geometric models defined below.

In [ ]:
test_acc = test()

out = model(data.x)
if enable_wandb:
    embedding_to_wandb(out, color=data.y, key="mlp/embedding/trained")
    wandb.summary["mlp/accuracy"] = test_acc
    wandb.log({"mlp/accuracy": test_acc})
else:
  visualize(out, data.y)

print(f'Test Accuracy: {test_acc:.4f}')

Test Accuracy: 0.6040


As one can see, our MLP performs rather bad with only about 59% test accuracy.
But why does the MLP do not perform better?
The main reason for that is that this model suffers from heavy overfitting due to only having access to a **small amount of training nodes**, and therefore generalizes poorly to unseen node representations.

It also fails to incorporate an important bias into the model: **Cited papers are very likely related to the category of a document**.
That is exactly where Graph Neural Networks come into play and can help to boost the performance of our model.



## Training a Graph Neural Network (GNN)

We can easily convert our MLP to a GNN by swapping the `torch.nn.Linear` layers with PyG's GNN operators.

Following-up on [the first part of this tutorial](https://colab.research.google.com/drive/1h3-vJGRVloF5zStxL5I0rSy4ZUPNsjy8), we replace the linear layers by the [`GCNConv`](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.GCNConv) module.
To recap, the **GCN layer** ([Kipf et al. (2017)](https://arxiv.org/abs/1609.02907)) is defined as

$$
\mathbf{x}_v^{(\ell + 1)} = \mathbf{W}^{(\ell + 1)} \sum_{w \in \mathcal{N}(v) \, \cup \, \{ v \}} \frac{1}{c_{w,v}} \cdot \mathbf{x}_w^{(\ell)}
$$

where $\mathbf{W}^{(\ell + 1)}$ denotes a trainable weight matrix of shape `[num_output_features, num_input_features]` and $c_{w,v}$ refers to a fixed normalization coefficient for each edge.
In contrast, a single `Linear` layer is defined as

$$
\mathbf{x}_v^{(\ell + 1)} = \mathbf{W}^{(\ell + 1)} \mathbf{x}_v^{(\ell)}
$$

which does not make use of neighboring node information.

In [ ]:
from torch_geometric.nn import GCNConv


class GCN(torch.nn.Module):
    def __init__(self, hidden_channels):
        super().__init__()
        torch.manual_seed(1234567)
        self.conv1 = GCNConv(dataset.num_features, hidden_channels)
        self.conv2 = GCNConv(hidden_channels, dataset.num_classes)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index)
        x = x.relu()
        x = F.dropout(x, p=0.5, training=self.training)
        x = self.conv2(x, edge_index)
        return x

model = GCN(hidden_channels=16)
print(model)

GCN(
  (conv1): GCNConv(1433, 16)
  (conv2): GCNConv(16, 7)
)


Let's visualize the node embeddings of our **untrained** GCN network.
For visualization, we make use of [**TSNE**](https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html) to embed our 7-dimensional node embeddings onto a 2D plane.

In [ ]:
model = GCN(hidden_channels=16)
model.eval()

out = model(data.x, data.edge_index)

if enable_wandb:
    embedding_to_wandb(out, color=data.y, key="gcn/embedding/init")
else:
    visualize(out, data.y)

We certainly can do better by training our model.
The training and testing procedure is once again the same, but this time we make use of the node features `x` **and** the graph connectivity `edge_index` as input to our GCN model.

In [ ]:
from IPython.display import Javascript  # Restrict height of output cell.
display(Javascript('''google.colab.output.setIframeHeight(0, true, {maxHeight: 300})'''))

model = GCN(hidden_channels=16)
if enable_wandb:
    wandb.watch(model)

optimizer = torch.optim.Adam(model.parameters(), lr=0.01, weight_decay=5e-4)
criterion = torch.nn.CrossEntropyLoss()

def train():
      model.train()
      optimizer.zero_grad()  # Clear gradients.
      out = model(data.x, data.edge_index)  # Perform a single forward pass.
      loss = criterion(out[data.train_mask], data.y[data.train_mask])  # Compute the loss solely based on the training nodes.
      loss.backward()  # Derive gradients.
      optimizer.step()  # Update parameters based on gradients.
      return loss

def test():
      model.eval()
      out = model(data.x, data.edge_index)
      pred = out.argmax(dim=1)  # Use the class with highest probability.
      test_correct = pred[data.test_mask] == data.y[data.test_mask]  # Check against ground-truth labels.
      test_acc = int(test_correct.sum()) / int(data.test_mask.sum())  # Derive ratio of correct predictions.
      return test_acc


for epoch in range(1, 101):
    loss = train()
    if enable_wandb:
        wandb.log({"gcn/loss": loss})
    print(f'Epoch: {epoch:03d}, Loss: {loss:.4f}')

<IPython.core.display.Javascript object>

Epoch: 001, Loss: 1.9463
Epoch: 002, Loss: 1.9409
Epoch: 003, Loss: 1.9343
Epoch: 004, Loss: 1.9275
Epoch: 005, Loss: 1.9181
Epoch: 006, Loss: 1.9086
Epoch: 007, Loss: 1.9015
Epoch: 008, Loss: 1.8933
Epoch: 009, Loss: 1.8808
Epoch: 010, Loss: 1.8685
Epoch: 011, Loss: 1.8598
Epoch: 012, Loss: 1.8482
Epoch: 013, Loss: 1.8290
Epoch: 014, Loss: 1.8233
Epoch: 015, Loss: 1.8057
Epoch: 016, Loss: 1.7966
Epoch: 017, Loss: 1.7825
Epoch: 018, Loss: 1.7617
Epoch: 019, Loss: 1.7491
Epoch: 020, Loss: 1.7310
Epoch: 021, Loss: 1.7147
Epoch: 022, Loss: 1.7056
Epoch: 023, Loss: 1.6954
Epoch: 024, Loss: 1.6697
Epoch: 025, Loss: 1.6538
Epoch: 026, Loss: 1.6312
Epoch: 027, Loss: 1.6161
Epoch: 028, Loss: 1.5899
Epoch: 029, Loss: 1.5711
Epoch: 030, Loss: 1.5576
Epoch: 031, Loss: 1.5393
Epoch: 032, Loss: 1.5137
Epoch: 033, Loss: 1.4948
Epoch: 034, Loss: 1.4913
Epoch: 035, Loss: 1.4698
Epoch: 036, Loss: 1.3998
Epoch: 037, Loss: 1.4041
Epoch: 038, Loss: 1.3761
Epoch: 039, Loss: 1.3631
Epoch: 040, Loss: 1.3258


After training the model, we can check its test accuracy:

In [ ]:
test_acc = test()
print(f'Test Accuracy: {test_acc:.4f}')

Test Accuracy: 0.8150


**There it is!**
By simply swapping the linear layers with GNN layers, we can reach **81.5% of test accuracy**!
This is in stark contrast to the 59% of test accuracy obtained by our MLP, indicating that relational information plays a crucial role in obtaining better performance.

We can also verify that once again by looking at the output embeddings of our **trained** model, which now produces a far better clustering of nodes of the same category.

In [ ]:
model.eval()

out = model(data.x, data.edge_index)

if enable_wandb:
    wandb.summary["gcn/accuracy"] = test_acc
    wandb.log({"gcn/accuracy": test_acc})
    embedding_to_wandb(out, color=data.y, key="gcn/embedding/trained")
    wandb.finish()
else:
    visualize(out, data.y)

## Using W&B Sweeps

In this section, we'll look into how we can use [W&B Sweeps](https://wandb.ai/site/sweeps/) to perform a hyper-parameter search for the GCN. For this to work, it is essential for wandb to be enabled, i.e., `enable_wandb` should be set to `True`.

In [ ]:
assert enable_wandb, "W&B not enabled. Please, enable W&B and restart the notebook"

In [ ]:
import tqdm

def agent_fn():
    wandb.init()
    model = GCN(hidden_channels=wandb.config.hidden_channels)
    wandb.watch(model)

    with torch.no_grad():
      out = model(data.x, data.edge_index)
      embedding_to_wandb(out, color=data.y, key="gcn/embedding/init")

    optimizer = torch.optim.Adam(model.parameters(), lr=wandb.config.lr, weight_decay=wandb.config.weight_decay)
    criterion = torch.nn.CrossEntropyLoss()

    def train():
          model.train()
          optimizer.zero_grad()  # Clear gradients.
          out = model(data.x, data.edge_index)  # Perform a single forward pass.
          loss = criterion(out[data.train_mask], data.y[data.train_mask])  # Compute the loss solely based on the training nodes.
          loss.backward()  # Derive gradients.
          optimizer.step()  # Update parameters based on gradients.
          return loss

    def test():
          model.eval()
          out = model(data.x, data.edge_index)
          pred = out.argmax(dim=1)  # Use the class with highest probability.
          test_correct = pred[data.test_mask] == data.y[data.test_mask]  # Check against ground-truth labels.
          test_acc = int(test_correct.sum()) / int(data.test_mask.sum())  # Derive ratio of correct predictions.
          return test_acc


    for epoch in tqdm.tqdm(range(1, 101)):
        loss = train()
        wandb.log({"gcn/loss": loss})


    model.eval()

    out = model(data.x, data.edge_index)
    test_acc = test()
    wandb.summary["gcn/accuracy"] = test_acc
    wandb.log({"gcn/accuracy": test_acc})
    embedding_to_wandb(out, color=data.y, key="gcn/embedding/trained")
    wandb.finish()

In [ ]:
sweep_config = {
    "name": "gcn-sweep",
    "method": "bayes",
    "metric": {
        "name": "gcn/accuracy",
        "goal": "maximize",
    },
    "parameters": {
        "hidden_channels": {
            "values": [8, 16, 32]
        },
        "weight_decay": {
            "distribution": "normal",
            "mu": 5e-4,
            "sigma": 1e-5,
        },
        "lr": {
            "min": 1e-4,
            "max": 1e-3
        }
    }
}

# Register the Sweep with W&B
sweep_id = wandb.sweep(sweep_config, project="node-classification")

Create sweep with ID: p1ktaxqt
Sweep URL: https://wandb.ai/morg/node-classification/sweeps/p1ktaxqt


In [ ]:
# Run the Sweeps agent
wandb.agent(sweep_id, project="node-classification", function=agent_fn, count=50)

wandb: Agent Starting Run: et1qb89e with config:
wandb: 	hidden_channels: 8
wandb: 	lr: 0.0009847987424734062
wandb: 	weight_decay: 0.0004931451386070596


100%|██████████| 100/100 [00:02<00:00, 40.72it/s]


wandb: Agent Starting Run: q0cs17cp with config:
wandb: 	hidden_channels: 8
wandb: 	lr: 0.0009901178498644455
wandb: 	weight_decay: 0.0004895866883954844


100%|██████████| 100/100 [00:02<00:00, 39.85it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: qrzsgidh with config:
wandb: 	hidden_channels: 8
wandb: 	lr: 0.0009932461685218112
wandb: 	weight_decay: 0.0004837818251903144


100%|██████████| 100/100 [00:02<00:00, 39.69it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: b3ai1e2a with config:
wandb: 	hidden_channels: 8
wandb: 	lr: 0.0009682879185034716
wandb: 	weight_decay: 0.0004849645392544168


100%|██████████| 100/100 [00:02<00:00, 38.98it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: la9gtb2z with config:
wandb: 	hidden_channels: 8
wandb: 	lr: 0.0009941452161455497
wandb: 	weight_decay: 0.0004898851853941055


100%|██████████| 100/100 [00:02<00:00, 47.14it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: eazc6qvc with config:
wandb: 	hidden_channels: 16
wandb: 	lr: 0.0009532420000025276
wandb: 	weight_decay: 0.000474478076341511


100%|██████████| 100/100 [00:02<00:00, 36.12it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: z3suirwf with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0008822590493801454
wandb: 	weight_decay: 0.0004852058160335215


100%|██████████| 100/100 [00:03<00:00, 31.69it/s]


wandb: Agent Starting Run: 1gct4o0y with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0008845486976276018
wandb: 	weight_decay: 0.0004883468015428194


100%|██████████| 100/100 [00:03<00:00, 29.71it/s]


wandb: Agent Starting Run: 165k5uui with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0008352731326967797
wandb: 	weight_decay: 0.0004845417998739997


100%|██████████| 100/100 [00:03<00:00, 28.37it/s]


wandb: Agent Starting Run: mw31x1m4 with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009583715103678278
wandb: 	weight_decay: 0.0004809804241144013


100%|██████████| 100/100 [00:03<00:00, 27.62it/s]


wandb: Agent Starting Run: lxks3zsd with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009685210113976614
wandb: 	weight_decay: 0.0004856507940324258


100%|██████████| 100/100 [00:03<00:00, 27.61it/s]


wandb: Agent Starting Run: uyy2xcli with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009760980390330432
wandb: 	weight_decay: 0.000482935914053525


100%|██████████| 100/100 [00:03<00:00, 31.07it/s]


wandb: Agent Starting Run: emlwy175 with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.000997507348957398
wandb: 	weight_decay: 0.00047948929480089245


100%|██████████| 100/100 [00:03<00:00, 28.13it/s]


wandb: Agent Starting Run: jb6n4vvr with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009878369593796638
wandb: 	weight_decay: 0.00048251969143776305


100%|██████████| 100/100 [00:03<00:00, 25.65it/s]


wandb: Agent Starting Run: 8jtx58c7 with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.000995177047063983
wandb: 	weight_decay: 0.0004821131480246045


100%|██████████| 100/100 [00:03<00:00, 27.15it/s]


wandb: Agent Starting Run: 3886m97x with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009589388300136788
wandb: 	weight_decay: 0.0004866753229429843


100%|██████████| 100/100 [00:03<00:00, 26.49it/s]


wandb: Agent Starting Run: meb74wiz with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009848735117822052
wandb: 	weight_decay: 0.00048625246409945966


100%|██████████| 100/100 [00:03<00:00, 27.60it/s]


wandb: Agent Starting Run: k08fqj1s with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009901042453669769
wandb: 	weight_decay: 0.0004837614472856642


100%|██████████| 100/100 [00:03<00:00, 29.64it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: okqbp8no with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.000998178438092641
wandb: 	weight_decay: 0.0004899039812065805


100%|██████████| 100/100 [00:03<00:00, 27.56it/s]


wandb: Agent Starting Run: yw6dfv1g with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009721271322580448
wandb: 	weight_decay: 0.0004857625760776506


100%|██████████| 100/100 [00:03<00:00, 26.49it/s]


wandb: Agent Starting Run: 02vpvimd with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009498340806644516
wandb: 	weight_decay: 0.0004874377678205295


100%|██████████| 100/100 [00:03<00:00, 28.17it/s]


wandb: Agent Starting Run: hylg76y3 with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009666729543825766
wandb: 	weight_decay: 0.0004863083750322432


100%|██████████| 100/100 [00:03<00:00, 29.50it/s]


wandb: Agent Starting Run: rttb0jkd with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009140198680451892
wandb: 	weight_decay: 0.0004909701512093212


100%|██████████| 100/100 [00:03<00:00, 28.92it/s]


wandb: Agent Starting Run: r2s3myrf with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009847346052650472
wandb: 	weight_decay: 0.000488786878667419


100%|██████████| 100/100 [00:03<00:00, 25.51it/s]


wandb: Agent Starting Run: m4wcu2hp with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009867068925499115
wandb: 	weight_decay: 0.0004895422319978684


100%|██████████| 100/100 [00:03<00:00, 29.81it/s]


wandb: Agent Starting Run: yk2cd2rc with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.000949360938724521
wandb: 	weight_decay: 0.0004893057997634927


100%|██████████| 100/100 [00:03<00:00, 29.67it/s]


wandb: Agent Starting Run: btmnr2dp with config:
wandb: 	hidden_channels: 16
wandb: 	lr: 0.0009239708409857672
wandb: 	weight_decay: 0.000485068357911382


100%|██████████| 100/100 [00:02<00:00, 36.17it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: 6ieg7cam with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009683295745297038
wandb: 	weight_decay: 0.0004783178919663449


100%|██████████| 100/100 [00:03<00:00, 28.36it/s]


wandb: Agent Starting Run: rhcrl6hy with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009090924720856502
wandb: 	weight_decay: 0.00048817251274983943


100%|██████████| 100/100 [00:03<00:00, 27.48it/s]


wandb: Agent Starting Run: wmozrtcu with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009913135768394995
wandb: 	weight_decay: 0.0004848423657979577


100%|██████████| 100/100 [00:03<00:00, 27.53it/s]


wandb: Agent Starting Run: xxbumgzr with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009383395067820352
wandb: 	weight_decay: 0.000473255765472336


100%|██████████| 100/100 [00:03<00:00, 25.29it/s]


wandb: Agent Starting Run: k5cp4d2u with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009962175343680478
wandb: 	weight_decay: 0.000491188709614926


100%|██████████| 100/100 [00:03<00:00, 28.25it/s]


wandb: Agent Starting Run: v15bi5jj with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009437398397553286
wandb: 	weight_decay: 0.0004786766477396835


100%|██████████| 100/100 [00:03<00:00, 28.79it/s]


wandb: Agent Starting Run: furnhcco with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009284072115972664
wandb: 	weight_decay: 0.00048492942709273395


100%|██████████| 100/100 [00:03<00:00, 29.36it/s]


wandb: Agent Starting Run: g7iezxf3 with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009237709257612524
wandb: 	weight_decay: 0.0004907993058301969


100%|██████████| 100/100 [00:03<00:00, 28.97it/s]


wandb: Agent Starting Run: hautjbdn with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009942758121193164
wandb: 	weight_decay: 0.0004877860165546661


100%|██████████| 100/100 [00:03<00:00, 30.02it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: lidirnu5 with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009071359660499352
wandb: 	weight_decay: 0.0004840978413155681


100%|██████████| 100/100 [00:03<00:00, 31.80it/s]


wandb: Agent Starting Run: nwgkodcv with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009925293205225393
wandb: 	weight_decay: 0.0004861973016127004


100%|██████████| 100/100 [00:03<00:00, 28.80it/s]


wandb: Agent Starting Run: 2m745lw9 with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009258388308979743
wandb: 	weight_decay: 0.0004841439470535139


100%|██████████| 100/100 [00:03<00:00, 28.92it/s]


wandb: Agent Starting Run: yu9deufk with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0009907296009142683
wandb: 	weight_decay: 0.00048385231393731777


100%|██████████| 100/100 [00:04<00:00, 23.11it/s]


wandb: Agent Starting Run: 86kg001x with config:
wandb: 	hidden_channels: 16
wandb: 	lr: 0.0003989920876165757
wandb: 	weight_decay: 0.0004885771013593518


100%|██████████| 100/100 [00:02<00:00, 37.63it/s]


wandb: Agent Starting Run: 9s1wo0b0 with config:
wandb: 	hidden_channels: 8
wandb: 	lr: 0.00048140869073364466
wandb: 	weight_decay: 0.0005105403266991152


100%|██████████| 100/100 [00:02<00:00, 44.77it/s]


wandb: Agent Starting Run: ndg2plxp with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.00013530341241974742
wandb: 	weight_decay: 0.0005075029380249654


100%|██████████| 100/100 [00:03<00:00, 29.46it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: fa23ssex with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0003660296752545047
wandb: 	weight_decay: 0.0005100239681278396


100%|██████████| 100/100 [00:03<00:00, 29.65it/s]


wandb: Agent Starting Run: 5y68bzt7 with config:
wandb: 	hidden_channels: 8
wandb: 	lr: 0.0007211280441718825
wandb: 	weight_decay: 0.00048372507978434217


100%|██████████| 100/100 [00:02<00:00, 35.96it/s]


wandb: Agent Starting Run: b3ctgv1c with config:
wandb: 	hidden_channels: 16
wandb: 	lr: 0.0005201492492172016
wandb: 	weight_decay: 0.0005139817816954003


100%|██████████| 100/100 [00:02<00:00, 34.76it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: bmo0252l with config:
wandb: 	hidden_channels: 8
wandb: 	lr: 0.0009087290087283952
wandb: 	weight_decay: 0.0004888262517100899


100%|██████████| 100/100 [00:02<00:00, 43.43it/s]


wandb: Agent Starting Run: a83llmud with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.0002583027105749223
wandb: 	weight_decay: 0.0005079775873928059


100%|██████████| 100/100 [00:03<00:00, 29.40it/s]


wandb: Agent Starting Run: 843wo4ba with config:
wandb: 	hidden_channels: 16
wandb: 	lr: 0.0007418890996066532
wandb: 	weight_decay: 0.0005027122723159877


100%|██████████| 100/100 [00:02<00:00, 36.64it/s]


wandb: Sweep Agent: Waiting for job.
wandb: Job received.
wandb: Agent Starting Run: t1yejmcr with config:
wandb: 	hidden_channels: 32
wandb: 	lr: 0.00047501254211357975
wandb: 	weight_decay: 0.0004918280636464704


100%|██████████| 100/100 [00:03<00:00, 29.77it/s]


## Conclusion

In this chapter, you have seen how to apply GNNs to real-world problems, and, in particular, how they can effectively be used for boosting a model's performance.
In the next section, we will look into how GNNs can be used for the task of graph classification.

[Next: Graph Classification with Graph Neural Networks](https://colab.research.google.com/drive/1I8a0DfQ3fI7Njc62__mVXUlcAleUclnb)

## (Optional) Exercises

1. To achieve better model performance and to avoid overfitting, it is usually a good idea to select the best model based on an additional validation set.
The `Cora` dataset provides a validation node set as `data.val_mask`, but we haven't used it yet.
Can you modify the code to select and test the model with the highest validation performance?
This should bring test performance to **82% accuracy**.

2. How does `GCN` behave when increasing the hidden feature dimensionality or the number of layers?
Does increasing the number of layers help at all?

3. You can try to use different GNN layers to see how model performance changes. What happens if you swap out all `GCNConv` instances with [`GATConv`](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.GATConv) layers that make use of attention? Try to write a 2-layer `GAT` model that makes use of 8 attention heads in the first layer and 1 attention head in the second layer, uses a `dropout` ratio of `0.6` inside and outside each `GATConv` call, and uses a `hidden_channels` dimensions of `8` per head.

In [ ]:
from torch_geometric.nn import GATConv


class GAT(torch.nn.Module):
    def __init__(self, hidden_channels, heads):
        super().__init__()
        torch.manual_seed(1234567)
        self.conv1 = GATConv(...)  # TODO
        self.conv2 = GATConv(...)  # TODO

    def forward(self, x, edge_index):
        x = F.dropout(x, p=0.6, training=self.training)
        x = self.conv1(x, edge_index)
        x = F.elu(x)
        x = F.dropout(x, p=0.6, training=self.training)
        x = self.conv2(x, edge_index)
        return x

model = GAT(hidden_channels=8, heads=8)
print(model)

optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=5e-4)
criterion = torch.nn.CrossEntropyLoss()

def train():
      model.train()
      optimizer.zero_grad()  # Clear gradients.
      out = model(data.x, data.edge_index)  # Perform a single forward pass.
      loss = criterion(out[data.train_mask], data.y[data.train_mask])  # Compute the loss solely based on the training nodes.
      loss.backward()  # Derive gradients.
      optimizer.step()  # Update parameters based on gradients.
      return loss

def test(mask):
      model.eval()
      out = model(data.x, data.edge_index)
      pred = out.argmax(dim=1)  # Use the class with highest probability.
      correct = pred[mask] == data.y[mask]  # Check against ground-truth labels.
      acc = int(correct.sum()) / int(mask.sum())  # Derive ratio of correct predictions.
      return acc


for epoch in range(1, 201):
    loss = train()
    val_acc = test(data.val_mask)
    test_acc = test(data.test_mask)
    print(f'Epoch: {epoch:03d}, Loss: {loss:.4f}, Val: {val_acc:.4f}, Test: {test_acc:.4f}')